### ISSUE TITLE
3.4 regression on type inference with union typed array spreads

### ISSUE THREAD

**Issue by mceachen** [microsoft/TypeScript#30720]
Reproduced with TS 3.4.1 and typescript@3.4.0-dev.20190330.

**Search Terms:**

Union typed array spreads used to work with TS 3.3 but require explicit types with TS 3.4.

**Code**

Works as expected on both 3.3 and 3.4:
```ts
declare function concat<T>(...args: (T | T[])[]): T[]
concat([1, 2, 3], 4)
```

Works on 3.3, breaks under 3.4 and @next:
```ts
export type Maybe<T> = T | undefined
declare function concatMaybe<T>(...args: (Maybe<T> | Maybe<T>[])[]): T[]

concatMaybe([1, 2, 3], 4)
```

**Unexpected error**

`error TS2345: Argument of type '4' is not assignable to parameter of type 'number[] | (number[] | undefined)[] | undefined'.`

This _does_ compile under 3.4, fwiw:

```ts
concatMaybe<number>([1, 2, 3], 4)
```

**Related Issues:** 

#30694

**Comment by mceachen** [comment]
@ahejlsberg I don't know if you and your team get this enough, but THANK YOU FOR TYPESCRIPT!

You and your team are doing great work. Kudos!

**Comment by ahejlsberg** [comment]
@mceachen Thanks, much appreciated!

**Comment by craigkovatch** [comment]
For real, TypeScript is the closest I have ever felt to expressing thought into code at the speed I can think it.
